src/browser-polyfill.js

```diff
@@ -6,7 +6,52 @@
  * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
 "use strict";
 
-if (typeof browser === "undefined" || Object.getPrototypeOf(browser) !== Object.prototype) {
+/**
+ * Returns true if the given object is an object with a `then` method, and can
+ * therefore be assumed to behave as a Promise.
+ *
+ * @param {*} value The value to test.
+ * @returns {boolean} True if the value is thenable.
+ */
+const isThenable = value => {
+  return value && typeof value === "object" && typeof value.then === "function";
+};
+
+/**
+ * Checks if the `browser` namespace object is an actual
+ * `browser` namespace object and not an element with
+ * the `[id]` attribute with a value of `browser`.
+ */
+const HAS_BROWSER_NAMESPACE = (typeof browser !== "undefined" && Object.getPrototypeOf(browser) === Object.prototype);
+
+/**
+ * Checks if the current browser supports
+ * the Promise-based WebExtension APIs.
+ *
+ * @returns {boolean} True if yes.
+ */
+const supportsPromises = () => {
+  if (!HAS_BROWSER_NAMESPACE) {
+    return false;
+  }
+  // If `browser.runtime.lastError` doesn’t exist, assume promises are supported.
```

Comment:
In Edge, these two checks don't work. The `browser.runtime.lastError` property is present, but it's `undefined` in the absence of an error. Maybe a check for `hasOwnProperty("lastError")` would be more appropriate, but I am not an expert.

However, `browser.extension.lastError` is not defined at all, even though Edge does not support promises and needs the polyfill.